### PR TITLE
Lineskip practice trails

### PR DESCRIPTION
**Pull request by MarkJeronimus**
A toggle is added to the 'automap' section of the options menu. (No key binding is added.)

When the full automap cheat is used, this feature shows a trail of the last 350 unique coordinates of the player on the map (at least 10 seconds of movement). This can greatly aid speedrunners in finding successful lineskip set-ups.

The image below shows a successful lineskip in MAP01, southwards, if that line had a trigger: one tic your center is still above the line, next tic your entire hitbox is below the line.

image

PS. I hope the "local player" implementation is correct in case of network play (I haven't tested this)

### CODE REVIEW

**Comment by MarkJeronimus** [comment]
Thoughts:
- Works best with map textures turned off. Textures are on by default, and this may confuse users who didn't rtfm.
- The option only works with IDDT×2 (to ensure 'strict' no-cheat compatibility when recording actual speedruns), this may also confuse players that expect it to "just work" when the option is enabled.
- When the option is enabled, the trail is calculated even when not visible via IDDT. This is my suggestion, not a requirement.

File: prboom2/src/m_menu.c [hunk 1]
```diff
@@ -3949,7 +3950,6 @@ setup_menu_t helpstrings[] =  // HELP screen strings
   {"STRAFE RIGHT",S_SKIP|S_INPUT,m_null,KT_X3,0,dsda_input_straferight},
   {"STRAFE"      ,S_SKIP|S_INPUT,m_null,KT_X3,0,dsda_input_strafe},
   {"AUTORUN"     ,S_SKIP|S_INPUT,m_null,KT_X3,0,dsda_input_autorun},
-  {"180 TURN"    ,S_SKIP|S_INPUT,m_null,KT_X3,0,dsda_input_reverse},
```
Comment (kraflab) [inline prboom2/src/m_menu.c:3952]: Why is this change here?

File: prboom2/src/g_game.c [hunk 2]
```diff
@@ -3065,6 +3066,7 @@ void G_InitNew(int skill, int episode, int map, dboolean prepare)
 
   //jff 4/16/98 force marks on automap cleared every new level start
   AM_clearMarks();
+  AM_clearTrail();
```
Comment (kraflab) [inline prboom2/src/g_game.c:3069]: Let's rather include the call to AM_clearTrail inside AM_clearMarks itself.

File: prboom2/src/m_menu.c [hunk 3]
```diff
@@ -2701,6 +2701,7 @@ setup_menu_t auto_settings1[] =  // 1st AutoMap Settings screen
   { "Enable textured display", S_YESNO, m_conf, AU_X, dsda_config_map_textured },
   { "Things appearance", S_CHOICE, m_conf, AU_X, dsda_config_map_things_appearance, 0, map_things_appearance_list },
   { "Show Minimap", S_YESNO, m_conf, AU_X, dsda_config_show_minimap },
+  { "Lineskip practice trails", S_YESNO, m_conf, AU_X, dsda_config_map_trail },
```
Comment (kraflab) [inline prboom2/src/m_menu.c:2704]: Let's call this "Player position trail" as it could be relevant for more things than just line skips.

File: prboom2/src/am_map.c [hunk 4]
```diff
@@ -1218,6 +1278,9 @@ void AM_Ticker (void)
 
   if (stop_zooming && leveltime - zoom_leveltime != 1)
     AM_StopZooming();
+
+  if (dsda_IntConfig(dsda_config_map_trail))
+    AM_updateTrail();
```
Comment (kraflab) [inline prboom2/src/am_map.c:1283]: I think this isn't quite exactly where we want this to happen. The critical action is when the player's position is updated, especially if it means they can trigger lines at that moment. That happens at the tail end of P_TryMove. It's not quite the same as a once per tick position check because high speeds cause multiple P_TryMove calls within a tic (this is why wallrunning works, for example).

**Review by kraflab — COMMENTED** [review]
Nice work! There are some small adjustments to be made but the feature is sound 👍

File: prboom2/src/m_menu.c (hunk 1 above)
Comment (MarkJeronimus) [reply prboom2/src/m_menu.c:3952]: Just something I stumbled upon while working with the sources. If there's a reason it should not be removed, then you can drop this commit (idk if that's how it works)

File: prboom2/src/g_game.c (hunk 2 above)
Comment (MarkJeronimus) [reply prboom2/src/g_game.c:3069]: I consider trails separate from marks

File: prboom2/src/m_menu.c (hunk 3 above)
Comment (MarkJeronimus) [reply prboom2/src/m_menu.c:2704]: Sounds good to me.

File: prboom2/src/am_map.c (hunk 4 above)
Comment (MarkJeronimus) [reply prboom2/src/am_map.c:1283]: Yup. I thought about split movements and moved it to the back of my mind, as something to tackle later, and then completely forgot about it.

**Comment by kraflab** [comment]
I've adopted the foundations of this PR here: https://github.com/kraflab/dsda-doom/commit/ec542d1837c2eee9aa4d5d9257631c49be31622f (basically the core code with some changes but not hooked up yet). I'm testing different integrations and will probably extend this to a few different modes (tracking all position checks vs checking locations where lines are checked etc). I have a video on the discord if you're curious how that looks with fences for instance. I'll hook things up soon so no need to update this PR, but I'll leave it up until the feature is available in the dev branch.

**Comment by MarkJeronimus** [comment]
Notes:
I made a typo (style error) in trailpoint_t. I indented one space instead of two. I like to keep in the style on whatever I'm working on and because this differs from my own style it wasn't obvious to me.

If you're drawing 'Cross marking center' and 'Bounding box' in the same color, it would make more sense to draw 'Bounding box' first instead of the other way around (which was only the case so the Bounding box was drawn on top)

I'll look for the discord. [edit] I can't find it.

**Comment by kraflab** [comment]
Finished for now https://github.com/kraflab/dsda-doom/commit/60e9c15ed857a334fc6f779beb9628b8ace4ac25

@MarkJeronimus https://discord.gg/yxaFP8T59H

**Comment by MarkJeronimus** [comment]
I found that invite before, but I was searching for DSDA, not Hell on Earth.

But the invite link says I'm banned there. Since I've never been there nor known anyone from there I can only assume it's whitelisted?
